### PR TITLE
Add support for local custom authorizers

### LINKED ISSUE #404: Local mode authorizers
Allow for authorizers to work in local mode for better testing.

### PR DESCRIPTION
**Pull request by stealthycoin**
BuiltinAuthorizer will now work in local mode. Authorizer types that
cannot be run locally will now simply allow the requests to proceed as
if there were no authorizer to allow for local testing.

closes #404

### CODE REVIEW

File: chalice/local.py [hunk 1]
```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+        # type: (str, int, int, Optional[Clock]) -> None
+        if time_source is None:
+            time_source = Clock()
+        self._time_source = time_source
+        self._start_time = self._current_time_millis()
+        self._max_runtime = runtime_millis
+        self._initialize_lambda_context_attributes(
+            function_name, memory_size)
+
+    def _initialize_lambda_context_attributes(self, function_name,
```
Comment (JordonPhillips) [inline chalice/local.py]: It's a bit bizzare to define instance attributes outside of init, especially since you're not doing any complicated initialization or logic to set their initial values.

File: chalice/local.py [hunk 2]
```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+        lambda_event['requestContext']['authorizer'] = auth_context
+        return lambda_event
 
-    def __init__(self, request, client_address, server, app_object):
-        # type: (bytes, Tuple[str, int], HTTPServer, Chalice) -> None
-        self.app_object = app_object
+    def _prepare_authorizer_event(self, lambda_event):
+        # type: (EventType) -> EventType
+        """Translate event for an authorizer input."""
+        authorizer_event = lambda_event.copy()
```
Comment (JordonPhillips) [inline chalice/local.py]: How nested is this structure? if it's more than one level then maybe you should deepcopy.

File: chalice/local.py [hunk 3]
```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+        # type: (EventType) -> Tuple[Optional[Any], Optional[RouteEntry]]
+        # Authorizer had to be made into an Any type since mypy couldn't
+        # detect that app.ChaliceAuthorizer was callable.
+        resource_path = lambda_event.get(
+            'requestContext', {}).get('resourcePath')
+        http_method = lambda_event['requestContext']['httpMethod']
+        try:
+            route_entry = self._app_object.routes[resource_path][http_method]
+        except KeyError:
+            # If there a key error is raised when trying to get the route entry
```
Comment (JordonPhillips) [inline chalice/local.py]: s/If there/If/

File: chalice/local.py [hunk 4]
```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+            # then this route does not support this method. A method error
+            # will be raised by the chalice handler method. We can ignore it
+            # here by returning no authorizer to avoid duplicating the logic.
+            return None, None
+        authorizer = route_entry.authorizer
+        return authorizer, route_entry
```
Comment (JordonPhillips) [inline chalice/local.py]: If `authorizer` is an attribute of `route_entry` why are you not just returning `route_entry`?

File: chalice/local.py [hunk 5]
```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+
+    def authorize(self, lambda_event, lambda_context):
+        # type: (EventType, LambdaContext) -> Tuple[EventType, LambdaContext]
+        auth_event = self._prepare_authorizer_event(lambda_event)
+        authorizer, route_entry = self._authorizer_and_route_for_event(
+            lambda_event)
+        # Technically we only need to check for authorizer being `None` here
+        # since the route_entry will always match authorizer but mypy needs us
+        # to explicitly check both before we can use them. In any event, if
+        # either of these are not defined then there was no authroizer or
```
Comment (JordonPhillips) [inline chalice/local.py]: You seem to have trailed off here.

File: chalice/local.py [hunk 6]
```diff
@@ -222,21 +394,99 @@ def _send_autogen_options_response(self, lambda_event):
...
+
+    def _handle_binary(self, response):
+        # type: (Dict[str,Any]) -> Dict[str,Any]
+        if response.get('isBase64Encoded'):
+            body = base64.b64decode(response['body'])
+            response['body'] = body
+        return response
+
+
+class ChaliceRequestHandler(BaseHTTPRequestHandler):
```
Comment (JordonPhillips) [inline chalice/local.py]: A class doc string for this would be nice

File: chalice/local.py (hunk 1 above)
Comment (stealthycoin) [inline chalice/local.py]: I separated the logic that is specific to the actual lambda context initialization vs the initializations I have to do for my fake object to run.

File: chalice/local.py (hunk 4 above)
Comment (stealthycoin) [inline chalice/local.py]: Yes I think this made more sense before a refactor and I didn't think to update it.

File: chalice/local.py (hunk 2 above)
Comment (stealthycoin) [inline chalice/local.py]: The event is a dict of entirely [str, Union[str, Dict[str, str]] so I don't think it needs to be deepcopied unless I am mistaken about something.

File: chalice/local.py [hunk 7]
```diff
@@ -109,98 +128,254 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+        method_pattern = r'(?:\*|%s)' % route_entry.method
+        route_entry_pattern = r'^.*?/%s%s%s$' % (method_pattern,
+                                                 route_entry.uri_pattern,
+                                                 pattern_suffix)
+        route_entry_re = re.compile(route_entry_pattern)
+        # Look through all statements for one that allows us to invoke this
+        # particular view function.
+        for statement in statements:
+            if statement.get('Effect') == 'Allow' and \
+               statement.get('Action') == 'execute-api:Invoke':
```
Comment (jamesls) [inline chalice/local.py]: These technically support wildcards as well.  If you're not going to support them, you should probably make a note about eventually supporting them.

File: chalice/local.py [hunk 8]
```diff
@@ -109,98 +128,254 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+        # user can return any policy at all from the custom authorizer.
+        pattern_suffix = ''
+        if route_entry.uri_pattern == '/':
+            # There is a special case when the / is used since the API Gateway
+            # arn will have the / duplicated. On all other routes it will
+            # omit the leading /.
+            pattern_suffix = '/'
+        method_pattern = r'(?:\*|%s)' % route_entry.method
+        route_entry_pattern = r'^.*?/%s%s%s$' % (method_pattern,
+                                                 route_entry.uri_pattern,
```
Comment (jamesls) [inline chalice/local.py]: I don't think this is right.  The uri pattern will include the placeholder values so something like this won't work:

```python
@app.route('/foo/{name}', authorizer=demo_auth)
def index(name):
    return {'context': app.current_request.context}

```

If I return `AuthResponse(routes=['/foo/bar'], principal_id='user')` from my authorizer.

What if we mirrored things more closely to real IAM? 

1. Generate the resource ARN for the specific method/view
2. Check if any of the resources in the returned policy match the ARN (accounting for wildcards).

I also would suggest extracting this policy matching out to a separate class so we can add a lot more tests for the edge cases (using `pytest.paramtrize`).

File: chalice/app.pyi [hunk 9]
```diff
@@ -1,3 +1,4 @@
+from chalice.local import LambdaContext
```
Comment (jamesls) [inline chalice/app.pyi]: Import orders, chalice after 3rd party libs.

File: chalice/local.py [hunk 10]
```diff
@@ -4,27 +4,99 @@
...
+        # type: () -> float
+        return time.time()
+
+
+def create_local_server(app_obj, config, port):
+    # type: (Chalice, Config, int) -> LocalDevServer
+    return LocalDevServer(app_obj, config, port)
+
+
+class LocalArnBuilder(object):
```
Comment (jamesls) [inline chalice/local.py]: s/Arn/ARN

File: chalice/local.py [hunk 11]
```diff
@@ -4,27 +4,99 @@
...
+            region=self.LOCAL_REGION,
+            account_id=self.LOCAL_ACCOUNT_ID,
+            api_id=self.LOCAL_API_ID,
+            stage=self.LOCAL_STAGE,
+            method=method,
+            resource_path=path
+        )
+
+
+class ArnMatcher(object):
```
Comment (jamesls) [inline chalice/local.py]: s/Arn/ARN

File: chalice/local.py [hunk 12]
```diff
@@ -109,98 +181,246 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+            # BuiltinAuthConfig type. Anything else we will err on the side of
+            # allowing local testing by simply admiting the request. Otherwise
+            # there is no way for users to test their code in local mode.
+            warnings.warn(
+                '%s is not a supported in local mode. All requests made '
+                'against a route will be authorized to allow local testing.'
+                % authorizer.__class__.__name__
+            )
+            return lambda_event, lambda_context
+        auth_result = authorizer(auth_event, lambda_context)
```
Comment (jamesls) [inline chalice/local.py]: I don't think this is quite right.  If you don't have the appropriate auth header (`Authorization`) in the request, you get a 401 and the authorizer isn't called at all.

File: chalice/local.py [hunk 13]
```diff
@@ -214,29 +434,106 @@ def _send_autogen_options_response(self, lambda_event):
...
+            )
+            self._send_http_response(response)
+        except NotAuthorizedError as e:
+            self._send_not_authorized_response(e.headers)
+        except NoOptionsRouteDefined as e:
+            self._send_http_response_no_body(e.headers)
+
+    def _send_not_authorized_response(self, headers):
+        # type: (HeaderType) -> None
+        self.send_response(401)
```
Comment (jamesls) [inline chalice/local.py]: An authorized error where we generate a policy that forbids access to a resource gives a 403, not a 401.

```
$ http https://rest-api-id.execute-api.us-west-2.amazonaws.com/api/ 'Authorization: deny'
HTTP/1.1 403 Forbidden
Connection: keep-alive
Content-Length: 60
Content-Type: application/json
Date: Thu, 24 Aug 2017 00:27:57 GMT
Via: 1.1 7f3f42df8af148df1f9f1ee7175987ad.cloudfront.net (CloudFront)
X-Amz-Cf-Id: 4ZhC3GokHeXr68EKakpF6xQWxVvTDZtQiGK253psOcWbPW3M980rGw==
X-Cache: Error from cloudfront
x-amzn-ErrorType: AccessDeniedException
x-amzn-RequestId: 13ff2f6d-8863-11e7-af78-e1cc9bcdaa69

{
    "Message": "User is not authorized to access this resource"
}
```

File: tests/unit/test_app.py [hunk 14]
```diff
@@ -8,6 +8,9 @@
 from chalice import app
 from chalice import NotFoundError
 from chalice import __version__ as chalice_version
+from tests.unit.conftest import create_event
```
Comment (jamesls) [inline tests/unit/test_app.py]: I don't think conftest is meant to be imported directly (or at least it's something I see).  I'd suggest converting them to fixtures if you want to move them there.

I think I might have missed this in the wheel based packager PR.

File: chalice/local.py [hunk 15]
```diff
@@ -4,27 +4,99 @@
...
+
+def create_local_server(app_obj, config, port):
+    # type: (Chalice, Config, int) -> LocalDevServer
+    return LocalDevServer(app_obj, config, port)
+
+
+class LocalArnBuilder(object):
+    ARN_FORMAT = ('arn:aws:execute-api:{region}:{account_id}'
+                  ':{api_id}/{stage}/{method}/{resource_path}')
+    LOCAL_REGION = 'mars-west-1'
```
Comment (jamesls) [inline chalice/local.py]: Just curious, why aren't we using a real region as a default for the local region?

File: chalice/local.py [hunk 16]
```diff
@@ -109,98 +181,246 @@ def create_lambda_event(self, method, path, headers, body=None):
...
+
+class LambdaContext(object):
+    def __init__(self, function_name, memory_size,
+                 runtime_millis=3000, time_source=None):
+        # type: (str, int, int, Optional[Clock]) -> None
+        if time_source is None:
+            time_source = Clock()
+        self._time_source = time_source
+        self._start_time = self._current_time_millis()
+        self._max_runtime = runtime_millis
```
Comment (jamesls) [inline chalice/local.py]: `runtime_millis` is pretty vague, but `max_runtime_ms` is a lot clearer.  Is there a reason to not map to call the arg something different than it's instance attribute name?

File: tests/unit/test_local.py [hunk 17]
```diff
@@ -404,5 +559,344 @@ def test_can_create_lambda_event_for_post_with_formencoded_body():
 
 
 def test_can_provide_port_to_local_server(sample_app):
-    dev_server = local.create_local_server(sample_app, port=23456)
+    dev_server = local.create_local_server(sample_app, None, port=23456)
     assert dev_server.server.server_port == 23456
+
+
+class FakeTimeSource(object):
```
Comment (jamesls) [inline tests/unit/test_local.py]: We usually put all the non test classes/functions at the top of the file before the actual tests.

File: tests/unit/test_local.py [hunk 18]
```diff
@@ -404,5 +559,344 @@ def test_can_create_lambda_event_for_post_with_formencoded_body():
...
+        # API Gateway arn generation where an extra / is appended to the end
+        # of the arn.
+        authorizer = LocalGatewayAuthorizer(demo_app_auth)
+        event = create_event('/', 'GET', {})
+        event['headers']['authorization'] = 'allow'
+        context = LambdaContext(*lambda_context_args)
+        event, context = authorizer.authorize(event, context)
+        assert event['requestContext']['authorizer']['principalId'] == 'user'
+
+    def test_can_authorize_with_lower_case_headers(self, lambda_context_args,
```
Comment (jamesls) [inline tests/unit/test_local.py]: Isn't this identical to the previous test `test_can_authorize_empty_path`?

File: tests/unit/test_local.py [hunk 19]
```diff
@@ -404,5 +559,344 @@ def test_can_create_lambda_event_for_post_with_formencoded_body():
...
+        def index_view():
+            return {}
+
+        authorizer = LocalGatewayAuthorizer(demo)
+        event = create_event('/index', 'GET', {})
+        context = LambdaContext(*lambda_context_args)
+        event, context = authorizer.authorize(event, context)
+        # Assert that when the authorizer.authorize is called and there is no
+        # authorizer defined for a particular route that it is a noop.
+        assert event == event
```
Comment (jamesls) [inline tests/unit/test_local.py]: ?

File: tests/unit/test_local.py [hunk 20]
```diff
@@ -404,5 +559,344 @@ def test_can_create_lambda_event_for_post_with_formencoded_body():
...
+        assert context == context
```
Comment (jamesls) [inline tests/unit/test_local.py]: ?

File: chalice/local.py (hunk 15 above)
Comment (stealthycoin) [inline chalice/local.py]: Wanted to make it very clear it wasn't a real ARN and to force it never to match with any real ARNs unless they explicitly don't care about the region (with a glob). I can change it if you have a reason you would prefer a real region.

File: chalice/local.py (hunk 16 above)
Comment (stealthycoin) [inline chalice/local.py]: I wanted to give the unit in the name is why its not a direct mirror. I agree with `max_runtime_ms` so I'll do that.

File: tests/unit/test_local.py (hunk 17 above)
Comment (stealthycoin) [inline tests/unit/test_local.py]: I was keeping it nearby for easier reference. I'll move it.

File: tests/unit/test_local.py (hunk 18 above)
Comment (stealthycoin) [inline tests/unit/test_local.py]: Yea...

File: tests/unit/test_local.py (hunk 19 above)
Comment (stealthycoin) [inline tests/unit/test_local.py]: Fixed... still passes.

**Comment by jamesls** [comment]
I'm just going through this now, but it looks like there's still some feedback that hasn't been incorporated: https://github.com/aws/chalice/pull/467#discussion_r132775015

The ARN generated for a route that has a captured param is still: `'arn:aws:execute-api:mars-west-1:123456789012:ymy8tbxw7b/api/GET/foo/{name}'` which isn't right.  I tested the example in the linked comment and it still fails in local mode but passes on actual API gateway/lambda.
